src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -1014,6 +1014,33 @@ private ValueTypeShapeCharacteristics ComputeHomogeneousAggregateCharacteristic(
             return NotHA;
         }
 
+        public override bool ComputeIsUnsafeValueType(DefType type)
+        {
+            if (!type.IsValueType)
+                return false;
+
+            MetadataType metadataType = (MetadataType)type;
+            if (metadataType.HasCustomAttribute("System.Runtime.CompilerServices", "UnsafeValueTypeAttribute"))
+                return true;
+
+            foreach (FieldDesc field in metadataType.GetFields())
+            {
+                if (field.IsStatic || field.IsLiteral)
+                    continue;
+
+                if (field.FieldType is not DefType fieldType)
+                    continue;
+
+                if (fieldType.IsPrimitive)
+                    continue;
+
+                if (fieldType.IsUnsafeValueType)
+                    return true;
```

Comment:
This will be slightly more efficient (avoid computing the bit for reference type fields, less casting):

```suggestion
                TypeDesc fieldType = field.FieldType;
                if (!fieldType.IsValueType || fieldType.IsPrimitive)
                    continue;

                if ((DefType)fieldType).IsUnsafeValueType)
                    return true;
```